sound/soc/intel/boards/cml_rt5682.c

```diff
@@ -0,0 +1,327 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright(c) 2018 Intel Corporation.
+
+/*
+ * Intel Cometlake I2S Machine Driver with RT5682 Codecs
+ *
+ * Modified from:
+ *   Intel Apollolake I2S Machine driver
+ */
+
+#include <linux/input.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <sound/core.h>
+#include <sound/jack.h>
+#include <sound/pcm.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include "../../codecs/rt5682.h"
+#include "../../codecs/hdac_hdmi.h"
+
+#define CML_PLAT_CLK_FREQ 24000000
+#define RT5682_PLL_FREQ (48000 * 512)
+#define DUAL_CHANNEL 2
+#define QUAD_CHANNEL 4
+#define NAME_SIZE 32
+
+static struct snd_soc_jack cometlake_hdmi[3];
+
+struct cml_hdmi_pcm {
+	struct list_head head;
+	struct snd_soc_dai *codec_dai;
+	int device;
+};
+
+struct cml_card_private {
+	struct snd_soc_jack cometlake_headset;
+	struct list_head hdmi_pcm_list;
+};
+
+enum {
+	CML_DPCM_AUDIO_PB = 0,
+	CML_DPCM_AUDIO_CP,
+	CML_DPCM_AUDIO_HS_PB,
+	CML_DPCM_AUDIO_ECHO_REF_CP,
+	CML_DPCM_AUDIO_REF_CP,
+	CML_DPCM_AUDIO_DMIC_CP,
+	CML_DPCM_AUDIO_HDMI1_PB,
+	CML_DPCM_AUDIO_HDMI2_PB,
+	CML_DPCM_AUDIO_HDMI3_PB,
+};
+
+static int cometlake_ssp_fixup(struct snd_soc_pcm_runtime *rtd,
+			struct snd_pcm_hw_params *params)
+{
+	struct snd_interval *rate = hw_param_interval(params,
+			SNDRV_PCM_HW_PARAM_RATE);
+	struct snd_interval *channels = hw_param_interval(params,
+			SNDRV_PCM_HW_PARAM_CHANNELS);
+	struct snd_mask *fmt = hw_param_mask(params, SNDRV_PCM_HW_PARAM_FORMAT);
+
+	/* The ADSP will convert the FE rate to 48k, stereo */
+	rate->min = rate->max = 48000;
+	channels->min = channels->max = DUAL_CHANNEL;
+
+	/* set SSP to 24 bit */
+	snd_mask_none(fmt);
+	snd_mask_set(fmt, SNDRV_PCM_FORMAT_S24_LE);
+
+	return 0;
+}
+
+static int cometlake_rt5682_hw_params(struct snd_pcm_substream *substream,
+	struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai = rtd->codec_dai;
+	int ret;
+
+	/* Set valid bitmask & configuration for I2S in 24 bit */
+	ret = snd_soc_dai_set_tdm_slot(codec_dai, 0x0, 0x0, 2, 24);
+	if (ret < 0) {
+		dev_err(rtd->dev, "set TDM slot err:%d\n", ret);
+		return ret;
+	}
+
```

Comment:
If we need to support different sampling rate, we will need to set pll here.

Reply:
this is useless for SOF.

Reply:
sorry, please ignore my comment, I was confused with some old SST code where the set_tdm_slot if used for the cpu_dai. This should be the same as in the existing glk_rt5682....c